File: bank-service/src/main/java/rs/edu/raf/mapper/ExchangeInvoiceMapper.java
package rs.edu.raf.mapper;

import lombok.AllArgsConstructor;
import org.springframework.stereotype.Component;
import rs.edu.raf.dto.CurrencyDTO;
import rs.edu.raf.dto.ExchangeInvoiceDTO;
import rs.edu.raf.dto.BankAccountDTO;
import rs.edu.raf.model.ExchangeInvoice;

@Component
@AllArgsConstructor
public class ExchangeInvoiceMapper {
    private BankAccountMapper bankAccountMapper;
    private CurrencyMapper currencyMapper;

    public ExchangeInvoiceDTO exchangeInvoiceToExchangeInvoiceDTO(ExchangeInvoice exchangeInvoice) {
        BankAccountDTO senderAccount = bankAccountMapper.bankAccountToBankAccountDTO(exchangeInvoice.getSenderAccount());
        BankAccountDTO receiverAccount = bankAccountMapper.bankAccountToBankAccountDTO(exchangeInvoice.getReceiverAccount());
        CurrencyDTO senderCurrency = currencyMapper.currencyToCurrencyDTO(exchangeInvoice.getSenderCurrency());
        CurrencyDTO receiverCurrency = currencyMapper.currencyToCurrencyDTO(exchangeInvoice.getReceiverCurrency());

        return new ExchangeInvoiceDTO(exchangeInvoice.getId(), senderAccount, receiverAccount,
                exchangeInvoice.getSenderAmount(), senderCurrency, receiverCurrency, exchangeInvoice.getExchangeRate(),
                exchangeInvoice.getProfit(), exchangeInvoice.getDateAndTime());
    }
}

File: bank-service-kotlin/src/main/kotlin/org/example/bankservicekotlin/mapper/ExchangeInvoiceMapper.kt
package org.example.bankservicekotlin.mapper

import org.example.bankservicekotlin.dto.BankAccountDTO
import org.example.bankservicekotlin.dto.CurrencyDTO
import org.example.bankservicekotlin.dto.ExchangeInvoiceDTO
import org.example.bankservicekotlin.model.ExchangeInvoice
import org.springframework.stereotype.Component

@Component
class ExchangeInvoiceMapper(
    private val bankAccountMapper: BankAccountMapper, private val currencyMapper: CurrencyMapper
) {

    fun exchangeInvoiceToExchangeInvoiceDTO(exchangeInvoice: ExchangeInvoice): ExchangeInvoiceDTO {
        val senderAccount: BankAccountDTO = bankAccountMapper.bankAccountToBankAccountDTO(exchangeInvoice.senderAccount)
        val receiverAccount: BankAccountDTO =
            bankAccountMapper.bankAccountToBankAccountDTO(exchangeInvoice.receiverAccount)
        val senderCurrency: CurrencyDTO = currencyMapper.currencyToCurrencyDTO(exchangeInvoice.senderCurrency)
        val receiverCurrency: CurrencyDTO = currencyMapper.currencyToCurrencyDTO(exchangeInvoice.receiverCurrency)

        return ExchangeInvoiceDTO(
            exchangeInvoice.id,
            senderAccount,
            receiverAccount,
            exchangeInvoice.senderAmount,
            senderCurrency,
            receiverCurrency,
            exchangeInvoice.exchangeRate,
            exchangeInvoice.profit,
            exchangeInvoice.dateAndTime
        )
    }
}